include/zio/zio_buf.h

```diff
@@ -0,0 +1,243 @@
+ * with a microcontroller focused RTOS in mind.
+ *
+ * As an example in the case of a 9 DoF IMU sensor where each channel
+ * produces a 16 bit value the buffer after each trigger without timestamps
+ * would contain 18 bytes of data, 22 bytes with the 32bit cycle counter
+ * timestamp.
+ */
+
+#ifndef ZEPHYR_INCLUDE_ZIO_BUF_H_
+#define ZEPHYR_INCLUDE_ZIO_BUF_H_
+
+#include <zephyr/types.h>
+#include <kernel.h>
+
+#include <zio/zio_dev.h>
+
+struct zio_buf;
+
+/**
+ * @brief Function to pull from the zio_buf a single datum
+ *
+ * The size of the datum is decided by the device and its set of active
+ * channels. Its expected the caller already knows the exact size of
+ * sample.
+ */
+typedef int (*zio_buf_pull_t)(struct zio_buf *buf, void *datum);
+
+/**
+ * @brief Function to setup a poll event for the zio_buf
+ *
+ * The device itself decides when data becomes available and how. Then in
+ * a zio application a k_poll_event is setup as a way of polling one or
+ * more buffers.
+ */
+typedef int (*zio_buf_poll_init_t)(struct zio_buf *buf,
+				   struct k_poll_event *evt);
+
+/**
+ * @brief Function to set watermark of zio_buf
+ *
+ * Optional function for a driver if watermark manipulation is possible
+ */
+typedef int (*zio_buf_set_watermark_t)(struct zio_buf *buf, u32_t watermark);
+
+/**
+ * @brief Function to get watermark of zio_buf
+ *
+ * Optional function for a driver to return the known watermark
+ */
+typedef int (*zio_buf_get_watermark_t)(struct zio_buf *buf, u32_t *watermark);
+
+/**
+ * @brief Function to get length of zio_buf
+ *
+ * Optional function for a driver to return the known length
+ */
+typedef int (*zio_buf_get_length_t)(struct zio_buf *buf, u32_t *length);
+
+
+/**
+ * @brief A pollable buffer interface for reading and writing to streams of data
+ *
+ * Most devices will want to simply use a fifo backed zio_buf but optionally
+ * devices may provide a hardware fifo backed zio_buf.
+ *
+ * The interface should support either. In the case of a fifo backed zio_buf
+ * devices should make a best effort to use DMA transfers rather than memcpy
+ */
+struct zio_buf_api {
+	zio_buf_pull_t pull;
+	zio_buf_poll_init_t poll_init;
+	zio_buf_set_watermark_t set_watermark;
+	zio_buf_get_watermark_t get_watermark;
+	zio_buf_get_length_t get_length;
+};
+
+/**
+ * @brief A pollable fifo-like buffer for reading and writing to streams of data
+ *
+ * An implementation should implement the api above statically optionally
+ * setting a void* to its own data to be used. Each device driver may
+ * implement their own zio_buf or base it on the software zio_fifo_buf
+ * implementation already provided.
+ */
+struct zio_buf {
+	bool circular : 1;
+	enum overflow_type {
+		ZIO_BUF_OVERFLOW_NONE,
+		ZIO_BUF_OVERFLOW_FLAG,
+		ZIO_BUF_OVERFLOW_COUNT,
+	} overflow_type;
+	u32_t overflow;
+	u32_t datum_size;
+	bool timestamps;
+	struct device *device;
+	struct zio_buf_api *buf_api;
+	void *buf_data;
```

Comment:
How memory for the buffer is allocated?

Reply:
buf_data is a private data pointer used by the api implementor. In the case of a fifo backed zio_buf it'd be a void * to the anonymous struct type created by the zio_fifo macros